src/python/gem5/resources/resource.py

```diff
@@ -972,6 +938,233 @@ def obtain_resource(
+        )
+        for resource_info in suite["workloads"]
+    ]
+    workload_json = get_multiple_resource_json_obj(db_query, clients)
+
+    # Creating the workload resource objects for each workload
+    # and setting the input group for each workload
+    workload_input_group_dict = {}
+    for workload in workload_json:
+        workload_input_group_dict[
+            _get_workload(
+                workload,
+                local_path,
+                resource_directory,
+                download_md5_mismatch,
+                clients,
+                gem5_version,
+                quiet,
+            )
+        ] = id_input_group_dict[workload["id"]]
+
+    suite["workloads"] = workload_input_group_dict
+    return SuiteResource(
+        local_path=local_path,
+        downloader=None,
+        **suite,
+    )
+
+
+def _get_workload(
+    workload: Dict[str, str],
+    local_path: str,
+    resource_directory: str,
+    download_md5_mismatch: bool,
+    clients: List[str],
+    gem5_version: str,
+    quiet: bool,
+) -> WorkloadResource:
+    """
+    :param workload: The workload JSON object.
+    :param local_path: The local path of the workload.
+    :param resource_directory: The resource directory.
+    :param download_md5_mismatch: If the resource is present, but does not have
+                                  the correct md5 value, the resource will be
+                                  deleted and re-downloaded if this value is ``True``.
+                                  Otherwise an exception will be thrown.
+    :param clients: A list of clients to search for the resource. If this
+                    parameter is not set, it will default search all clients.
+    :param gem5_version: The gem5 version to use to filter incompatible
+                         resource versions. By default set to the current gem5
+                         version.
+    :param quiet: If ``True``, suppress output. ``False`` by default.
+    """
+    params = {}
+    resources = workload["resources"]
+    # Mapping IDs to parameter names
+    id_to_param_map = {
+        resource_info["id"]: param
+        for param, resource_info in resources.items()
+    }
+
+    db_query = []
+    for resource in workload["resources"].values():
+        db_query.append(
+            ClientQuery(
+                resource_id=resource["id"],
+                resource_version=resource["resource_version"],
+                gem5_version=gem5_version,
+            )
+        )
+    # Fetching resources as a list of dicts
+    resource_details_list = get_multiple_resource_json_obj(db_query, clients)
+
+    # Preparing the final mapping of parameters to their complete JSON objects
+    param_to_resource_json = {}
+    for resource in resource_details_list:
+        resource_id = resource["id"]
+        if resource_id in id_to_param_map:
+            param = id_to_param_map[resource_id]
+            param_to_resource_json[param] = resource
```

Comment:
So, why not `workload["resources"]`
```suggestion
    for param_name, param_resource in workload["resources"].items():
    
        # This could probably be a fancy one-liner thing but I can't be bothered figuring it out.
        resource_match = None
        for resource in resource_details_list:
            if param_resource["id"] == resource["id"] && param_resource["version"] == resource["version"]:
                resource_match = resource
                break
        if not resource_match:
             raise Exception("error...")
        param_to_resource_json[param_name]=resource_match
```

Reply:
I think could also merge this into the for loop below and just have one for loop constructing these resources.